miniconf/src/option.rs

```diff
@@ -1,81 +1,113 @@
-use crate::{Error, Keys, Metadata, TreeDeserialize, TreeKey, TreeSerialize};
+use core::any::Any;
+
+use crate::{Error, Keys, Metadata, Traversal, TreeAny, TreeDeserialize, TreeKey, TreeSerialize};
 use serde::{de::Deserialize, Deserializer, Serialize, Serializer};
 
 // `Option` does not add to the path hierarchy (does not consume from `keys` or call `func`).
 // But it does add one Tree API layer between its `Tree<Y>` level
 // and its inner type `Tree<Y'>` level: `Y' = Y - 1`.
+// Otherwise e would not be able to distinguish between an augmented `Option<T>: TreeKey<0>`
+// and a plain-serde `Option<T>: Serialize/Deserialize/Any` in situations where the trait to
+// use is implicit (e.g. in an array). Also the bounds heuristics in the
+// derive macros assume that a field type `#[tree(depth=Y)] F<T>` calls its generic types at
+// `TreeKey<{Y - 1}>`. The latter could be ameliorated with a `bounds` derive macro attribute.
+
+fn get<'a, const Y: bool, T, K: Keys>(
+    opt: &'a Option<T>,
+    keys: &mut K,
+) -> Result<&'a T, Traversal> {
+    if Y {
+        keys.finalize::<0>()?;
+    }
+    opt.as_ref().ok_or(Traversal::Absent(0))
+}
+
+fn get_mut<'a, const Y: bool, T, K: Keys>(
+    opt: &'a mut Option<T>,
+    keys: &mut K,
+) -> Result<&'a mut T, Traversal> {
+    if Y {
+        keys.finalize::<0>()?;
+    }
+    opt.as_mut().ok_or(Traversal::Absent(0))
+}
 
 // the Y >= 2 cases:
 macro_rules! depth {
     ($($y:literal)+) => {$(
         impl<T: TreeKey<{$y - 1}>> TreeKey<$y> for Option<T> {
             fn len() -> usize {
-                0
+                unreachable!()
```

Comment:
Because anything we can return here is invalid (we don't do paths at this level, not even "zero paths") and I would like to ensure that if this is ever called, it crashes instead of using invalid data. We could make them fallible or move them to another non-pub trait. Maybe the latter is something that should/could be done.

Reply:
I think the sealed trait might be the way to go, ref https://predr.ag/blog/definitive-guide-to-sealed-traits-in-rust/

Reply:
spawned #208